User Service CRUD

taesukang-dev opened the pull request:
## 📌Linked Issues
- #19 

## ✏Change Details
- 간단한 CRUD 를 구현 및 테스트 하였습니다.
- UserDto, UserRequest 를 구현하였습니다.
- UserRequest 를 validation 체크하기 위해 spring validation 을 추가하였습니다.

## 💬Comment
- session, spring security 를 위해 작업중 pr 을 올립니다.

## 📑References
- 없음

## ✅Check List
- [x] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?

taesukang-dev committed:
[#19] config: spring validation 추가

validation을 위해 추가합니다.

taesukang-dev committed:
[#19] feat: 간단한 user service crud 구현

user service 를 위한 UserRequest, UserDto 를 포함한 Service 와 service test 를 추가했습니다.

taesukang-dev committed:
[#19] user repository 함수 추가

user service 를 위해 find by email, find by username 을 추가하고 테스트를 진행하였습니다.

sungyujeon commented on src/main/java/com/imhero/user/dto/request/UserRequest.java:
```diff
@@ -0,0 +1,24 @@
+    @Pattern(regexp="(?=.*[0-9])(?=.*[a-zA-Z])(?=.*\\W)(?=\\S+$).{8,20}",
+            message = "비밀번호는 영문 대,소문자와 숫자, 특수기호가 적어도 1개 이상씩 포함된 8자 ~ 20자의 비밀번호여야 합니다.")
+    private String password;
+    private String passwordCheck;
```
password check를 서버에서 한 번 더 수행하는 것에 대해서 다시 이야기해보면 좋을 것 같습니다.

password와 password를 확인하는 두 값을 같이 관리하는 이유는 제가 생각하기에 '사용자가 의도한 값을 정확히 입력했는가'를 검증하기 위함입니다. 사용자가 의도한 비밀번호를 정확하게 입력했는지는 사용자의 입력을 받는 client side에서 관심이 있는 것이라고 생각합니다.

반면에 서버에서는 사용자가 입력한 비밀번호를 추가 처리(암호화 등)하여서 저장소에 저장한 뒤에 로그인 등에서 조회하여 사용하는 것에 관심이 있습니다. 사용자가 의도한 값을 정확히 입력했는지 보다는, 이미 패스워드가 넘어온 시점에 해당 값은 사용자가 의도한 값이라는 전제가 있다고 봅니다.

도메인 로직(예를 들어 영문, 숫자, 특수기호를 조합한 비밀번호인가 등)이 아니라면 'request value 자체를 사용자가 의도한대로 정확히 입력했는가?' 라는 관점은 불필요한 것 같습니다.

taesukang-dev replied:
@sungyujeon 
리뷰 내용 반영되었습니다!

sungyujeon commented on src/test/groovy/com/imhero/user/service/UserServiceTest.groovy:
```diff
@@ -0,0 +1,216 @@
+        UserRepository userRepository = Mock(UserRepository.class)
+        UserService userService = new UserService(userRepository)
+        UserRequest userRequest = new UserRequest(email, password, passwordCheck, username)
+        User user = User.of(email, password, username, "N")
+
+        when:
+        userService.save(userRequest)
+
+        then:
+        userRepository.findUserByUsername(_) >> Optional.of(user)
```
```suggestion
        User user = User.of(email, password, username, "N")
        userRepository.findUserByEmail(_) >> Optional.of(user)

        when:
        userService.save(userRequest)

        then:
```

findByEmail을 통해서 기존에 가입한 회원인지를 확인한 뒤에 throw해야 테스트를 통과할 것 같습니다.

taesukang-dev replied:
@sungyujeon 
리뷰 내용 반영되었습니다!

sungyujeon commented on src/main/java/com/imhero/user/service/UserService.java:
```diff
@@ -0,0 +1,73 @@
+    public boolean withdraw(String email) {
+        if (StringUtils.trimAllWhitespace(email).length() == 0) {
+            throw new IllegalArgumentException("잘못된 요청입니다.");
+        }
```
UserRequest에서 validation 하고 있으므로 불필요한 로직으로 보입니다.

(추가)
이메일 형식에 맞는 데이터만 persistent data로 저장할텐데, domain에서 검증하는 것은 어떻게 생각하실까요? 현재 validation을 UserRequest dto에서 처리하고 있는데, 도메인 로직(이메일은 이메일 형식에 맞아야하고, 비밀번호는 영문/숫자/특수문자 등이 있어야 하고, 이름은 필수로 입력되어야 하는 등)으로 빼서 validation 검사를 하는 것은 어떨까요

taesukang-dev replied:
@sungyujeon 
리뷰 내용 반영되었습니다!

taesukang-dev replied:
@sungyujeon 
제 생각은 controller 단에서 validating 을 하는 것이 좋다는 것인데요, 단순히 형식에 부합하지 않을 때에 에러를 뱉어내는 것이 여러 레이어를 오고 가야할까 하는 생각입니다. service / repository 레이어 까지 내려온다면 하지 않아도 되는 애플리케이션의 리소스를 더 사용하게 될 것이라는 생각입니다.

sungyujeon commented on src/main/java/com/imhero/user/service/UserService.java:
```diff
@@ -0,0 +1,73 @@
+            throw new IllegalArgumentException("잘못된 요청입니다.");
+        }
+
+        User user = getUserByEmailOrElseThrow(email);
```
email을 파라미터로 받아서 user를 조회한 뒤에 탈회시키고 있는데, API spec에 보면 /users/{user_id}로 delete 요청을 보냅니다.

위 api를 그대로 가져간다면
1. user_id가 존재하는 사용자인지?
2. 존재하는 사용자라면 현재 세션에 로그인된 사용자만 탈퇴시키게 할 것인지?
등의 로직을 검증 후에 user_id를 기준으로 user를 찾아서 탈퇴시키게 하는 것이 될 것 같습니다.

한편으로는 특정 user_id를 가진 사용자가 해당 사용자의 정보를 update/delete 시키는 것은 권한의 문제일 수 있을 듯한데, api spec을 기존처럼 하는 것이 맞을지 다시 논의해보면 좋지 않을까 싶습니다. 예를 들어 /users/{user_id} put, delete 요청은 현재 로그인한 유저의 user_id와 일치할 때만 처리하는 것도 가능하지 않을까 싶습니다.

taesukang-dev replied:
@sungyujeon 
맞는 말씀이신 것 같습니다. 최초에 rest api / jwt 통신을 가정으로 하면서 나왔던 spec 으로 기억합니다. session 방식으로 바꾸면서 수정할 필요가 있을 것 같습니다!

taesukang-dev committed:
[#19] fix: 리뷰 내용 수정

UserRequest 의 passwordCheck 를 제외하였습니다.

taesukang-dev committed:
[#19] fix: service 에서 체크하는 email 검증 제거

UserRequest 에서 수정하고 있으므로 불필요하다고 판단하여 user service 에서 체크하는 email 공백 여부를 제거하였습니다.

taesukang-dev committed:
[#19] fix: 리뷰 내용 반영에 다른 테스트케이스 수정

UserRequest 에서 passwordCheck 를 제외함으로 변경되는 테스트케이스들을 수정하였습니다.

taesukang-dev committed:
[#19] fix: service 에서 체크하는 email 검증 제거

UserRequest 에서 검증하므로 리뷰 내용을 바탕으로 로직을 제외하였습니다.

taesukang-dev committed:
[#19] fix: 리뷰 내용 수정에 따른 테스트케이스 제거

service 에서 공백 체크를 하지 않음으로 제거하였습니다.

f-lab-joseph commented on src/main/java/com/imhero/user/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+public class UserService {
+    private final UserRepository userRepository;
+
+    // TODO : password encoder setting
```
TODO ?

taesukang-dev replied:
@f-lab-joseph 
주석 삭제하였습니다...!

f-lab-joseph commented on src/main/java/com/imhero/user/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+
+    // TODO : password encoder setting
+    // TODO : session setting
+    public UserDto save(UserRequest userRequest) {
```
transactional ?

taesukang-dev replied:
@f-lab-joseph 
서비스에 default 로 잡혀있습니다!

f-lab-joseph commented on src/main/java/com/imhero/user/service/UserService.java:
```diff
@@ -0,0 +1,62 @@
+    // TODO : session setting
+    public UserDto save(UserRequest userRequest) {
+        // TODO : Unique Constraint
+        userRepository.findUserByEmail(userRequest.getEmail()).ifPresent(it -> {
```
동시에 동일한 계정으로 가입요청이 들어가면 중복제거를 잘 해줄 수 있을까요 ?

taesukang-dev replied:
@f-lab-joseph 
말씀해주신대로 잘 해줄 수 없다고 생각했습니다.
따라서 해당 로직은 삭제하였습니다.
또한 user domain, email 컬럼에  unique key 를 걸어주었고  db 에 save 할 때 만약 이미 존재하는 email 일 경우 DataIntegrityViolationException 을 반환할 것이기 때문에 catch 로 잡아주는 방향으로 수정하였습니다.
감사합니다!

taesukang-dev committed:
[#19] fix: email 컬럼 unique key 추가

중복된 아이디로의 가입을 방지하고자 unique key 를 걸었습니다.
repository 에 관련된 테스트를 추가하였습니다.

taesukang-dev committed:
[#19] fix: 중복 이메일 가입 방지 로직 수정

중복 이메일을 체크하면서 동시성 문제가 발생할 수 있다고 판단하여 수정합니다.
findUserByEmail 을 제거하고 db 에 이미 unique key 가 걸려있으니 중복 시 DataIntegrityViolationException 을 잡아서 반환합니다.

taesukang-dev committed:
[#19] fix: ddl 수정

users email Unique key 추가하였습니다.

taesukang-dev committed:
[#19] fix: ddl 수정

users email Unique key 추가하였습니다.
